hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconContainerManager.java

```diff
@@ -77,22 +82,42 @@ public ReconContainerManager(
    * @throws IOException on Error.
    */
   public void checkAndAddNewContainer(ContainerID containerID,
+      ContainerReplicaProto.State replicaState,
       DatanodeDetails datanodeDetails)
       throws IOException {
     if (!exists(containerID)) {
       LOG.info("New container {} got from {}.", containerID,
           datanodeDetails.getHostName());
       ContainerWithPipeline containerWithPipeline =
           scmClient.getContainerWithPipeline(containerID.getId());
-      LOG.debug("Verified new container from SCM {} ",
-          containerWithPipeline.getContainerInfo().containerID());
+      LOG.info("Verified new container from SCM {}, part of {} ",
+          containerID, containerWithPipeline.getPipeline().getId());
       // If no other client added this, go ahead and add this container.
       if (!exists(containerID)) {
         addNewContainer(containerID.getId(), containerWithPipeline);
       }
+    } else {
+      // Check if container state is not open. In SCM, container state
+      // changes to CLOSING first, and then the close command is pushed down
+      // to Datanodes. Recon 'learns' this from DN, and hence replica state
+      // will move container state to 'CLOSING'.
+      ContainerInfo containerInfo = getContainer(containerID);
+      if (containerInfo.getState().equals(HddsProtos.LifeCycleState.OPEN)
+          && !replicaState.equals(ContainerReplicaProto.State.OPEN)
```

Comment:
should we more preciously check for ContainerReplicaProto.State.CLOSE state?

Reply:
@xiaoyuyao The container could be in any non-open & healthy state (CLOSING, QUASI_CLOSED, CLOSED). When the container was OPEN Recon may have added it to its metadata, but then it may have gone down. When it comes back up, the container could be in any state. A unit test (testProcessICRStateMismatch ) has been added for all the states.

Reply:
A sample flow of events.

- Recon has containerState = OPEN, Recon gets container report with replicaState = QUASI_CLOSED
- Before processing the replica, it calls on the above logic (ReconContainerManager#checkAndAddNewContainer) which updates containerState to ‘CLOSING’ through the "FINALIZE" event. This is like a pre-processing step for Recon where the container replica state is showing up as unexpected. We have to do this since SCM is where the state transition was initiated, but Recon has to "learn" it later from DNs. 
- Further along in the same process container report flow, there is a process replica step which sees that the containerState = ‘CLOSING’, and replicaState = QUASI_CLOSED, hence moves container state to ‘QUASI_CLOSED’ (AbstractContainerReportHandler#updateContainerState). This is the same code as used by SCM to transition container state by looking at the replica state.

The same sequence of events would be true if starting replicaState = “CLOSING" or "CLOSED” as well

Reply:
Thanks for the details of state transition handling on Recon side. It makes sense to me.